File: backend/app/services/survey_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.models.survey import SurveyResponse
from app.schemas.course import SurveySubmit
from app.utils.week_helpers import get_current_week_number, get_week_start_date
from app.utils.db_helpers import create_and_refresh
import uuid


async def submit_survey(db: AsyncSession, student_id: uuid.UUID, data: SurveySubmit):
    week = get_current_week_number()
    existing = await db.execute(
        select(SurveyResponse).where(
            SurveyResponse.student_id == student_id,
            SurveyResponse.survey_week == week,
        )
    )
    if existing.scalar_one_or_none():
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail="Already submitted survey for this week")

    week_start = get_week_start_date(week)
    response = SurveyResponse(
        student_id=student_id,
        survey_week=week,
        week_start_date=week_start,
        **data.model_dump(),
    )
    return await create_and_refresh(db, response)


async def get_student_surveys(db: AsyncSession, student_id: uuid.UUID):
    result = await db.execute(
        select(SurveyResponse)
        .where(SurveyResponse.student_id == student_id)
        .order_by(SurveyResponse.survey_week)
    )
    return result.scalars().all()


async def get_all_surveys(db: AsyncSession):
    result = await db.execute(select(SurveyResponse).order_by(SurveyResponse.survey_week))
    return result.scalars().all()


async def check_submitted(db: AsyncSession, student_id: uuid.UUID, week: int) -> bool:
    result = await db.execute(
        select(SurveyResponse).where(
            SurveyResponse.student_id == student_id,
            SurveyResponse.survey_week == week,
        )
    )
    return result.scalar_one_or_none() is not None

File: backend-php/src/Services/SurveyService.php
<?php
declare(strict_types=1);

namespace App\Services;

use App\Core\Database;
use App\Core\Config;

class SurveyService
{
    public static function getCurrentWeekNumber(): int
    {
        $studyStart = new \DateTime(Config::getInstance()->get('study_start_date', '2024-09-01'));
        $today      = new \DateTime('today');
        $delta      = $studyStart->diff($today)->days;
        return max(1, (int)($delta / 7) + 1);
    }

    public static function getWeekStartDate(int $weekNumber): string
    {
        $studyStart = new \DateTime(Config::getInstance()->get('study_start_date', '2024-09-01'));
        $studyStart->modify('+' . (($weekNumber - 1) * 7) . ' days');
        return $studyStart->format('Y-m-d');
    }

    public static function submitSurvey(string $studentId, array $data): array
    {
        $week = self::getCurrentWeekNumber();
        $pdo  = Database::getConnection();

        $stmt = $pdo->prepare('SELECT id FROM survey_responses WHERE student_id = ? AND survey_week = ?');
        $stmt->execute([$studentId, $week]);
        if ($stmt->fetch()) {
            throw new \RuntimeException('Already submitted survey for this week', 400);
        }

        $weekStart = self::getWeekStartDate($week);
        $stmt = $pdo->prepare('
            INSERT INTO survey_responses
                (student_id, survey_week, week_start_date, q1_punctuality_rating, q2_missed_classes,
                 q3_reminder_helpful, q4_location_preference, q5_privacy_comfort, q6_open_feedback)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            RETURNING *
        ');
        $stmt->execute([
            $studentId,
            $week,
            $weekStart,
            $data['q1_punctuality_rating'],
            $data['q2_missed_classes'],
            $data['q3_reminder_helpful'] ? 'TRUE' : 'FALSE',
            $data['q4_location_preference'],
            $data['q5_privacy_comfort'],
            $data['q6_open_feedback'] ?? null,
        ]);
        return $stmt->fetch();
    }

    public static function getStudentSurveys(string $studentId): array
    {
        $pdo  = Database::getConnection();
        $stmt = $pdo->prepare('SELECT * FROM survey_responses WHERE student_id = ? ORDER BY survey_week');
        $stmt->execute([$studentId]);
        return $stmt->fetchAll();
    }

    public static function getAllSurveys(): array
    {
        $pdo = Database::getConnection();
        return $pdo->query('SELECT * FROM survey_responses ORDER BY survey_week')->fetchAll();
    }

    public static function checkSubmitted(string $studentId, int $week): bool
    {
        $pdo  = Database::getConnection();
        $stmt = $pdo->prepare('SELECT id FROM survey_responses WHERE student_id = ? AND survey_week = ?');
        $stmt->execute([$studentId, $week]);
        return (bool)$stmt->fetch();
    }
}
